### PR TITLE
Add Ferret-core

### PR DESCRIPTION
**Pull request by xiangxiecrypto**
1. Add the core building blocks of Ferret (without network IO).
2. It can still be further optimized when integrated with network IO, which needs COT implementation.

### CODE REVIEW

**Comment by xiangxiecrypto** [comment]
Let me merge `MPCOT` and update accordingly.

**Comment by sinui0** [comment]
#87 is merged, you should be able to rebase

File: ot/mpz-ot-core/src/ferret/msgs.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+use serde::{Deserialize, Serialize};
+
+/// A Ferret message.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[allow(missing_docs)]
+pub enum Message<CotMsg, MpcotMsg> {
+    CotMsg(CotMsg),
+    MpcotMsg(MpcotMsg),
+    LpnMatrixSeed(LpnMatrixSeed),
+}
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/msgs.rs]: We can discard this now, as we're redesigning our approach for IO. See #112 . Changes there removes the need for a wrapper enum

File: ot/mpz-ot-core/src/ferret/sender.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    /// * `delta` - The sender's global secret.
+    /// * `lpn_parameters` - The lpn parameters.
+    /// * `lpn_type` - The lpn type.
+    /// * `seed` - The seed received from receiver to generate lpn matrix.
+    /// * `v` - The vector received from the COT ideal functionality.
+    pub fn setup(
+        self,
+        delta: Block,
+        lpn_parameters: LpnParameters,
+        lpn_type: LpnType,
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/sender.rs:38]: Can we move these to `new`?

File: ot/mpz-ot-core/src/ferret/mod.rs [hunk 3]
```diff
@@ -12,3 +18,131 @@ pub const CUCKOO_HASH_NUM: usize = 3;
...
+
+/// LPN parameters with regular noise.
+/// Derived from https://github.com/emp-toolkit/emp-ot/blob/master/emp-ot/ferret/constants.h
+pub const LPN_PARAMETERS_REGULAR: LpnParameters = LpnParameters {
+    n: 10180608,
+    k: 124000,
+    t: 4971,
+};
+
+/// LPN parameters with unifom noise.
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/mod.rs]:
```suggestion
/// LPN parameters with uniform noise.
```

File: ot/mpz-ot-core/src/ferret/sender.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        let mut y = s.to_vec();
+        self.state.lpn_encoder.compute(&mut y, &self.state.v);
+
+        // Update v
+        self.state.v = y[0..self.state.lpn_parameters.k].to_vec();
+
+        // Update counter
+        self.state.counter += 1;
+
+        Ok(y[self.state.lpn_parameters.k..].to_vec())
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/sender.rs]:
```suggestion
        let y_ = y.split_off(self.state.lpn_parameters.k);

        // Update v to y[0..k]
        self.state.v = y;

        // Update counter
        self.state.counter += 1;

        Ok(y_)
```

File: ot/mpz-ot-core/src/ferret/receiver.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    /// * `lpn_parameters` - The lpn parameters.
+    /// * `seed` - The seed to generate lpn matrix.
+    /// * `lpn_type` - The lpn type.
+    /// * `u` - The bits received from the COT ideal functionality.
+    /// * `w` - The vector received from the COT ideal functionality.
+    pub fn setup(
+        self,
+        lpn_parameters: LpnParameters,
+        lpn_type: LpnType,
+        seed: Block,
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/receiver.rs:40]: These can be moved to `new` as well

File: ot/mpz-ot-core/src/ferret/receiver.rs [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        self.state.lpn_encoder.compute(&mut z, &self.state.w);
+
+        // Compute x = A * u + e.
+        let u_block = self
+            .state
+            .u
+            .iter()
+            .map(|x| {
+                if *x {
+                    bytemuck::cast(1_u128)
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/receiver.rs]:
```suggestion
                    Block::ONE
```

Please add an associated const to `Block` so we don't rely on casts in places like this

File: ot/mpz-ot-core/src/ferret/receiver.rs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+                } else {
+                    Block::ZERO
+                }
+            })
+            .collect::<Vec<Block>>();
+        let mut x = self.state.e.to_vec();
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/receiver.rs]:
```suggestion
        let mut x = self.state.e.clone();
```

File: ot/mpz-ot-core/src/ferret/receiver.rs [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+        self.state.u = x[0..self.state.lpn_parameters.k].to_vec();
+
+        self.state.w = z[0..self.state.lpn_parameters.k].to_vec();
+
+        // Update counter
+        self.state.counter += 1;
+        Ok((
+            x[self.state.lpn_parameters.k..].to_vec(),
+            z[self.state.lpn_parameters.k..].to_vec(),
+        ))
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ferret/receiver.rs]:
```suggestion
        let mut x = x.iter().map(|a| a.lsb() == 1).collect::<Vec<bool>>();

        let x_ = x.split_off(self.state.lpn_parameters.k);
        let z_ = z.split_off(self.state.lpn_parameters.k);

        // Update u, w
        self.state.u = x;
        self.state.w = z;

        // Update counter
        self.state.counter += 1;

        Ok((x_, z_))
```

File: ot/mpz-ot-core/src/ideal/ideal_mpcot.rs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+    /// Performs the checks.
+    ///
+    /// # Arguments
+    ///
+    /// * `sender_msg` - The message that the ideal MPCOT sends to the sender.
+    /// * `receiver_msg` - The message that the ideal MPCOT sends to the receiver.
+    /// * `alphas` - The positions in each extension.
+    /// * `n` - The length of the vector.
+    pub fn check(
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ideal/ideal_mpcot.rs:85]: Please pull out into a standalone function gated with `#[cfg(test)]`

File: ot/mpz-ot-core/src/ideal/mod.rs [hunk 10]
```diff
@@ -1,4 +1,6 @@
 //! Define ideal functionalities of OTs.
+//! Define ideal functionalities of OTs.
```
Comment (sinui0) [inline ot/mpz-ot-core/src/ideal/mod.rs]: duplicate line

**Review by sinui0 — CHANGES_REQUESTED** [review]
Nice! Just a couple minor adjustments needed, otherwise lgtm :)

File: ot/mpz-ot-core/src/ferret/sender.rs (hunk 2 above)
Comment (xiangxiecrypto) [reply ot/mpz-ot-core/src/ferret/sender.rs:38]: Let us keep it here for now, the `new` function is only a default instantiation without any specific parameters.

**Review by sinui0 — APPROVED** [review]

**Pull request merged**